src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c

```diff
@@ -153,26 +175,56 @@ ucc_tl_ucp_reduce_scatter_knomial_progress(ucc_coll_task_t *coll_task)
                      (avg_pre_op ? ucc_knomial_pattern_loop_first_iteration(p)
                                  : ucc_knomial_pattern_loop_last_iteration(p));
 
-            status = ucc_tl_ucp_reduce_multi(
+            status = ucc_tl_ucp_reduce_multi_nb(
                 local_data, rbuf, reduce_data,
                 task->tagged.send_posted - p->iteration * (radix - 1),
-                local_seg_count, local_seg_count * dt_size, dt, mem_type, task,
-                is_avg);
+                local_seg_count, local_seg_count * dt_size, dt, task,
+                is_avg, exec, &task->reduce_scatter_kn.etask);
             if (ucc_unlikely(UCC_OK != status)) {
                 tl_error(UCC_TASK_LIB(task), "failed to perform dt reduction");
                 task->super.super.status = status;
                 return status;
             }
+UCC_KN_PHASE_REDUCE:
+            if (task->reduce_scatter_kn.etask != NULL) {
+                status = ucc_ee_executor_task_test(task->reduce_scatter_kn.etask);
+                if (status == UCC_INPROGRESS) {
+                    SAVE_STATE(UCC_KN_PHASE_REDUCE);
+                    return task->super.super.status;
+                }
+                ucc_ee_executor_task_finalize(task->reduce_scatter_kn.etask);
+                if (ucc_unlikely(status < 0)) {
+                    tl_error(UCC_TASK_LIB(task),
+                             "failed to perform dt reduction");
+                    task->super.super.status = status;
+                }
+            }
         }
         ucc_knomial_pattern_next_iteration(p);
     }
 
-    offset = ucc_sra_kn_get_offset(count, dt_size, rank, size, radix);
-    status = ucc_mc_memcpy(PTR_OFFSET(args->dst.info.buffer, offset),
-                           task->reduce_scatter_kn.scratch,
-                           local_seg_count * dt_size, mem_type, mem_type);
-
-    if (UCC_OK != status) {
+    ucc_sra_kn_get_offset_and_seglen(count, dt_size, rank, size, radix,
+                                     &offset, &local_seg_count);
+    eargs.task_type = UCC_EE_EXECUTOR_TASK_TYPE_COPY;
```

Comment:
all this maybe also can be put into a common macro/fn: tl_ucp_memcpy. It is also used in allgather.